src/control/lib/hostlist/hostlist.go

```diff
@@ -0,0 +1,751 @@
+//
+// (C) Copyright 2019 Intel Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// GOVERNMENT LICENSE RIGHTS-OPEN SOURCE SOFTWARE
+// The Government's rights to use, modify, reproduce, release, perform, display,
+// or disclose this software are subject to the terms of the Apache License as
+// provided in Contract No. 8F-30005.
+// Any reproduction of computer software, computer software documentation, or
+// portions thereof marked with this legend must also reproduce the markings.
+//
+package hostlist
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"sync"
+)
+
+var (
+	ErrEmpty    = errors.New("hostlist is empty")
+	ErrNotFound = errors.New("hostname not found")
+)
+
+const (
+	// MaxRange is the largest host range (hi - lo) supported
+	MaxRange = 16384
+)
+
+const (
+	outerRangeSeparators = "\t, "
+	innerRangeSeparator  = ","
+	rangeOperator        = "-"
+)
+
+type (
+	hostName struct {
+		prefix    string
+		suffix    string
+		number    uint
+		width     int
+		hasNumber bool
+	}
+
+	// HostList is a list of hostnames optimized for a "prefixXXXX"
+	// naming convention, where XXXX is a decimal, numeric suffix.
+	HostList struct {
+		sync.RWMutex
+		ranges    hostRanges
+		hostCount int
+	}
+)
+
+func (hn *hostName) Parse(input string) error {
+	// prefixN (default)
+	re := regexp.MustCompile(`^([a-zA-Z]+)(\d+)?(.*)?`)
+	if strings.Contains(input, "-") {
+		// handle hosts in the format prefixN-N
+		re = regexp.MustCompile(`^(\w+-?)(\d+)?(.*)?`)
+	}
```

Comment:
Want to cover any other hostname formats? What about something like prefix-N?

Reply:
Well, prefix-N would be covered by the second format, but I added a testcase for it to be sure. Are there other formats you think it should handle? I wanted to follow the LLNL C implementation pretty closely to start with.

Reply:
Nothing off the top of my head. prefix-N just seemed like a plausible one.